### PR TITLE
[Deprecated] Enable Indexer cache for DS v3.2 decoding

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

Enable Indexer cache for DS v3.2 decoding, to unblock the eval benchmark for DS v3.2 model with sparse attention bringup.
* DS reference implementation for Indexer is [here](https://github.com/deepseek-ai/DeepSeek-V3.2-Exp/blob/main/inference/model.py#L435-L495)
* Add `init_indexer_cache` & `update_indexer_cache` for indexer cache
* Other small changes, including formatting to pass the test (needed for `src/maxtext/layers/attention_op.py`) - sorry I have to rebase to resolve the conflict

# Tests

* All runners are green
* Training end-to-end (no impact) with a smaller model version: [link](https://paste.googleplex.com/4878875365605376)
* Test against reference implementation still green: [link](https://paste.googleplex.com/5361978873872384)
* Decoding gives reasonable 
  * small seq len to skip the indexer: [link](https://screenshot.googleplex.com/6hPKqZmCz47jUNZ)
  * large seq len to process the indexer `max_prefill_predict_length=3072 max_target_length=4096`: [link](https://screenshot.googleplex.com/5H9uqCovQ5d8vMy)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by shuningjin — COMMENTED** [review]
Thank you! There were some attention autoregressive/decoding tests:
- mha: https://github.com/AI-Hypercomputer/maxtext/blob/3fbe3be9830943857b6ded2f031588c85afaaea5/tests/unit/attention_test.py#L371
- mla: https://github.com/AI-Hypercomputer/maxtext/blob/3fbe3be9830943857b6ded2f031588c85afaaea5/tests/unit/attention_test.py#L1195

Shall we add a similar one to test indexer cache in mla? [example from gemini](https://paste.googleplex.com/5902350570094592)

**Comment by RissyRan** [comment]
> Thank you! There were some attention autoregressive/decoding tests:
> 
> * mha: https://github.com/AI-Hypercomputer/maxtext/blob/3fbe3be9830943857b6ded2f031588c85afaaea5/tests/unit/attention_test.py#L371
> * mla: https://github.com/AI-Hypercomputer/maxtext/blob/3fbe3be9830943857b6ded2f031588c85afaaea5/tests/unit/attention_test.py#L1195
> 
> Shall we add a similar one to test indexer cache in mla? [example from gemini](https://paste.googleplex.com/5902350570094592)

Sounds good. Also enabled the mla assertion, and did some sanity check for decoding: [long seq](https://screenshot.googleplex.com/3JNUxh32MXmBboW), [short seq](https://screenshot.googleplex.com/4evUZuChVnZJmJf)

File: src/maxtext/layers/attention_mla.py [hunk 1]
```diff
@@ -167,6 +176,31 @@ def __init__(
...
     )
 
+  def update_indexer_cache(self, kv_cache, k, decoder_segment_ids, model_mode, previous_chunk):
+    """Updates Indexer buffers by processing KV cache results."""
+    k_expanded = k[:, :, jnp.newaxis, :]
+    p_res, a_res = kv_cache(
+        key=k_expanded,
+        value=k_expanded,
+        decoder_segment_ids=decoder_segment_ids,
+        model_mode=model_mode,
```
Comment (Rohan-Bierneni) [inline src/maxtext/layers/attention_mla.py:186]: IIUC, a property of the indexer is that it only requires storage for the keys. 

However, we are initializing the indexer cache via KVCache, which stores keys and values. Passing `k_expanded` as the value doubles the HBM footprint. Is it possible to pass in a dummy value/tensor. Maybe of shape [B, S, 1, 1] for the value. 

Moreover, to accomodate the usage of KVCache for indexer class, we artificially unsqueeze `k` to 4D to satisfy the generic API, only to squeeze it back to 3D upon retrieval. This operation is redundant but would still work functionally.

Maybe a comment we can put here as a TODO is if a dedicated SparseKVCache class were written explicitly for the Indexer, we can drop the head dimension entirely and simplify the computation graph.

File: src/maxtext/layers/attention_mla.py [hunk 2]
```diff
@@ -615,16 +672,47 @@ def __init__(
...
     # Module attribute names must match names previously passed to Linen for checkpointing
     self.MlaKVCache_0 = self.init_mla_kv_caches(inputs_kv_shape) if model_mode != MODEL_MODE_TRAIN else None
 
+  def init_indexer_cache(self, inputs_kv_shape: Tuple):
+    """Initializes Indexer Cache."""
+    batch_size, _, _ = inputs_kv_shape
+    # Use standard KVCache to store keys. Values are unused but required by KVCache API.
+    # KVCache expects key_heads and value_heads. Since k is shared (MQA-like for Indexer),
+    # we use key_heads=1, value_heads=1.
+    return kvcache.KVCache(
```
Comment (Rohan-Bierneni) [inline src/maxtext/layers/attention_mla.py:695]: Looking at `MLA.__init__`, it sets up a `PagedAttentionOp` for the main query, key, and value computations if paged attention is enabled. However, the Indexer cache explicitly relies on the standard, dense kvcache.KVCache which would be incompatible for paged attention.

We should add a comment that paged attention is not supported with Sparse Attention.

File: src/maxtext/layers/attention_mla.py [hunk 3]
```diff
@@ -276,6 +314,16 @@ def __call__(
...
     k = k.squeeze(2)  # [b, s, 1, d] -> [b, s, d]
 
+    # Update and retrieve from cache if not training
+    cached_s = None
+    if model_mode != MODEL_MODE_TRAIN:
+      k_cached, cached_s = self.update_indexer_cache(kv_cache, k, decoder_segment_ids, model_mode, previous_chunk)
+      k = k_cached if k_cached is not None else k
+
+    # NOTE: If the total available sequence length <= topk, indexer always selects all tokens.
+    if k.shape[1] <= self.index_topk:
```
Comment (Rohan-Bierneni) [inline src/maxtext/layers/attention_mla.py:324]: For this check, doesn't k.shape[1] always equal to max_prefill_predict_length. Since JAX uses static shapes unlike Pytorch, then this condition would evaluate to False anytime max_prefill_predict_length > top_k value. 

However, the ideal behavior would be as decoding runs, the condition should check against the current token number that increases by 1 every decode step. 

I asked gemini to generate jax code for this check and gave this:
```
# 1. Dynamically count active tokens
# cached_s has > 0 for valid tokens, 0 for padding
active_token_count = jnp.sum(cached_s > 0, axis=-1) 

# 2. Define the heavy compute path
def compute_sparse_mask():
    logits = jnp.einsum("bthd, bsd -> btsh", q, k)
    # ... (rest of the heavy einsum and top_k math) ...
    return index_mask, topk_indices, index_score

# 3. Define the bypass path (Dense Attention)
def bypass_sparse_mask():
    # If we have fewer tokens than top-k, we just want standard dense attention.
    # Dense attention is achieved by simply returning the padding mask!
    index_mask = jnp.where(cached_s > 0, 0.0, DEFAULT_MASK_VALUE)
    
    # Return dummy tensors for the unused outputs to keep JAX happy
    dummy_topk = jnp.zeros((bsz, seqlen, self.index_topk), dtype=jnp.int32)
    dummy_score = jnp.zeros((bsz, seqlen, k.shape[1]), dtype=self.dtype)
    return index_mask, dummy_topk, dummy_score

# 4. Use jax.lax.cond to execute only ONE of the paths
index_mask, topk_indices, index_score = jax.lax.cond(
    (active_token_count > self.index_topk).any(), # Condition
    compute_sparse_mask,                          # Runs if True
    bypass_sparse_mask                            # Runs if False
)
```

**Review by Rohan-Bierneni — COMMENTED** [review]
Thank you for this PR! I have added a few comments PTAL.

File: src/maxtext/layers/attention_mla.py (hunk 3 above)
Comment (shuningjin) [reply src/maxtext/layers/attention_mla.py:324]: We could focus on functionality now, and optimize as a followup. Currently, the logic seems correct to me
```
train, k.shape[1] = max_target_length
prefill, k.shape[1] = max_prefill_length
autoregressive, k.shape[1] = max_target_length
```

e.g., I tried
```
python3 -m maxtext.inference.decode src/maxtext/configs/base.yml \
model_name=deepseek3.2-671b \
override_model_config=true base_num_decoder_layers=4 first_num_dense_layers=2 \
attention=dot_product scan_layers=false sparse_matmul=false \
dtype=bfloat16 weight_dtype=bfloat16 \
per_device_batch_size=1 max_prefill_predict_length=512 max_target_length=1024 mla_naive_kvcache=false 'prompt=An attention function can be described as mapping a query and a set of key-value pairs to an output, where the query, keys, values, and outputs are all vectors. The output is '
```
log shows `train: 1024, prefill: 512, autoregressive: 1024`

File: src/maxtext/layers/attention_mla.py (hunk 1 above)
Comment (shuningjin) [reply src/maxtext/layers/attention_mla.py:186]: We could focus on functionality now, and optimize as a followup, either via maxtext or vLLM.

Similar KV cache optimization: In the other PR which set value = key, we could similarly reduce KV cache to K cache only.  Their solution is to defer the decoding optimization to vLLM: https://github.com/AI-Hypercomputer/maxtext/pull/3283#issuecomment-3994572628

File: src/maxtext/layers/attention_mla.py (hunk 2 above)
Comment (RissyRan) [reply src/maxtext/layers/attention_mla.py:695]: Yeah, I think this is included in this code block? For paged attention, we need to set `attention=paged` to enable it.

```
    if self.use_indexer:
      if self.q_lora_rank == 0:
        raise NotImplementedError("Sparse indexer has not implemented for q_lora_rank = 0.")
      supports_dot_product = self.attention == "dot_product"
      supports_flash_splash = self.attention == "flash" and self.use_tokamax_splash
      if not (supports_dot_product or supports_flash_splash):
        raise NotImplementedError(
            "Sparse indexer is only supported dot_product attention or flash attention with tokamax splash."
        )
```

File: src/maxtext/layers/attention_mla.py (hunk 1 above)
Comment (RissyRan) [reply src/maxtext/layers/attention_mla.py:186]: Yeah, let's make it functional first. We could collaborate with inference team and make it more efficient.

File: src/maxtext/layers/attention_mla.py (hunk 3 above)
Comment (RissyRan) [reply src/maxtext/layers/attention_mla.py:324]: Please let me know if still has questions or concerns.
